File: ios/BundleDropBundleVerifier.swift
import Foundation

enum BundleDropBundleVerifier {
  private static let manifestFileName = "bundle-manifest.json"

  static func verifyBundleFiles(
    bundleDir: URL,
    manifestURL: URL,
    fileManager: FileManager = .default
  ) throws {
    let manifest: [String: Any]
    do {
      let data = try Data(contentsOf: manifestURL)
      guard let parsed = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
        throw BundleDropFileOps.error("Malformed bundle manifest")
      }
      manifest = parsed
    } catch {
      throw BundleDropFileOps.error("Malformed bundle manifest")
    }

    guard let files = manifest["files"] as? [[String: Any]] else {
      throw error("Bundle manifest must include files")
    }
    try verifyManifestFiles(bundleDir: bundleDir, files: files, fileManager: fileManager)
  }

  static func verifyManifestFiles(
    bundleDir: URL,
    files: [[String: Any]],
    fileManager: FileManager = .default
  ) throws {
    var allowed = Set([manifestFileName])

    for (index, file) in files.enumerated() {
      guard let path = file["path"] as? String else {
        throw error("Invalid manifest file entry at index \(index)")
      }
      guard isSafeManifestPath(path) else {
        throw error("Invalid manifest path: \(path)")
      }
      guard allowed.insert(path).inserted else {
        throw error("Duplicate manifest file path: \(path)")
      }
      guard let size = file["size"] as? NSNumber,
            let expectedSha = file["sha256"] as? String else {
        throw error("Invalid manifest file entry for \(path)")
      }

      let fileURL = bundleDir.appendingPathComponent(path)
      var isDirectory: ObjCBool = false
      guard fileManager.fileExists(atPath: fileURL.path, isDirectory: &isDirectory) else {
        throw error("Manifest file missing: \(path)")
      }
      guard !isDirectory.boolValue else {
        throw error("Manifest file is a directory: \(path)")
      }
      guard let attrs = try? fileManager.attributesOfItem(atPath: fileURL.path),
            let actualSize = attrs[.size] as? NSNumber,
            actualSize.int64Value == size.int64Value else {
        throw error("Manifest file size mismatch for \(path)")
      }
      guard try BundleDropFileOps.sha256File(fileURL.path) == expectedSha else {
        throw error("Manifest file hash mismatch for \(path)")
      }
    }

    let actualFiles = listRelativeFiles(root: bundleDir, fileManager: fileManager)
    if let extraFile = actualFiles.first(where: { !allowed.contains($0) }) {
      throw error("Unmanifested file in bundle archive: \(extraFile)")
    }
  }

  static func isSafeManifestPath(_ path: String) -> Bool {
    !path.isEmpty &&
      !path.hasPrefix("/") &&
      !path.contains("\\") &&
      !path.contains("\u{0000}") &&
      !path.split(separator: "/", omittingEmptySubsequences: false).contains { part in
        part.isEmpty || part == "." || part == ".."
      }
  }

  private static func listRelativeFiles(
    root: URL,
    current: URL? = nil,
    prefix: String = "",
    fileManager: FileManager
  ) -> [String] {
    let directory = current ?? root
    guard let entries = try? fileManager.contentsOfDirectory(at: directory, includingPropertiesForKeys: [.isDirectoryKey]) else {
      return []
    }
    return entries.flatMap { entry -> [String] in
      let relativePath = prefix.isEmpty ? entry.lastPathComponent : "\(prefix)/\(entry.lastPathComponent)"
      let isDirectory = (try? entry.resourceValues(forKeys: [.isDirectoryKey]).isDirectory) ?? false
      return isDirectory
        ? listRelativeFiles(root: root, current: entry, prefix: relativePath, fileManager: fileManager)
        : [relativePath]
    }
  }

  private static func error(_ message: String) -> NSError {
    BundleDropFileOps.error(message)
  }
}

File: android/src/main/java/com/bundledrop/BundleDropBundleVerifier.kt
package com.bundledrop

import org.json.JSONObject
import java.io.File
import java.io.IOException

object BundleDropBundleVerifier {
  private const val MANIFEST_FILE = "bundle-manifest.json"

  fun verifyBundleFiles(bundleDir: File, manifestFile: File) {
    val manifest = try {
      JSONObject(manifestFile.readText(Charsets.UTF_8))
    } catch (e: Exception) {
      throw IOException("Malformed bundle manifest", e)
    }
    val files = manifest.optJSONArray("files")
      ?: throw IOException("Bundle manifest must include files")
    verifyManifestFiles(bundleDir, files)
  }

  fun verifyManifestFiles(bundleDir: File, files: org.json.JSONArray) {
    val allowed = mutableSetOf(MANIFEST_FILE)

    for (i in 0 until files.length()) {
      val file = files.optJSONObject(i)
        ?: throw IOException("Invalid manifest file entry at index $i")
      val relativePath = file.optString("path", "")
      if (!isSafeManifestPath(relativePath)) {
        throw SecurityException("Invalid manifest path: $relativePath")
      }
      if (!allowed.add(relativePath)) {
        throw IOException("Duplicate manifest file path: $relativePath")
      }

      val expectedSize = file.optLong("size", -1L)
      val expectedSha = file.optString("sha256", "")
      val fullPath = File(bundleDir, relativePath)
      if (!fullPath.exists()) {
        throw IOException("Manifest file missing: $relativePath")
      }
      if (fullPath.isDirectory) {
        throw IOException("Manifest file is a directory: $relativePath")
      }
      if (fullPath.length() != expectedSize) {
        throw IOException("Manifest file size mismatch for $relativePath")
      }
      if (BundleDropFileOps.sha256File(fullPath) != expectedSha) {
        throw IOException("Manifest file hash mismatch for $relativePath")
      }
    }

    val actualFiles = listRelativeFiles(bundleDir)
    val extraFile = actualFiles.firstOrNull { !allowed.contains(it) }
    if (extraFile != null) {
      throw IOException("Unmanifested file in bundle archive: $extraFile")
    }
  }

  fun isSafeManifestPath(path: String): Boolean {
    return path.isNotBlank() &&
      !path.startsWith("/") &&
      !path.contains("\\") &&
      !path.contains("\u0000") &&
      path.split("/").none { it.isBlank() || it == "." || it == ".." }
  }

  private fun listRelativeFiles(root: File, dir: File = root, prefix: String = ""): List<String> {
    return dir.listFiles()?.flatMap { file ->
      val relativePath = if (prefix.isBlank()) file.name else "$prefix/${file.name}"
      if (file.isDirectory) listRelativeFiles(root, file, relativePath) else listOf(relativePath)
    } ?: emptyList()
  }
}
